app/models/access_control_upload.rb

```diff
@@ -0,0 +1,463 @@
+    roles.each do |item|
+      role = Role.where(name: item[:name]).first_or_create!
+      permissions = item[:permissions].map { |column, row| [column, row[:incoming_value]] }.to_h
+      role.update(**permissions)
+    end
+  end
+
+  private def import_agencies!
+    agencies.select { |_k, v| v[:existing_agency_id].blank? }.each_key do |name|
+      Agency.where(name: name).first_or_create!
+    end
+  end
+
+  # Add new users, update agencies and names for existing users, set everyone to using ACLs
+  private def import_users!
+    agency_ids = Agency.pluck(:name, :id).to_h
+    users.each do |item|
+      user = User.with_deleted.where(email: item[:email]).first_or_initialize do |u|
+        u.password = Faker::Internet.password(min_length: 16)
+      end
+      user.first_name = item[:first_name]
+      user.last_name = item[:last_name]
+      user.agency_id = agency_ids[item[:agency]]
+      user.permission_context = 'acls'
+      user.restore if user.deleted?
+      user.save!
+    end
+  end
+
+  private def import_user_groups!
+    user_groups.values.each do |item|
+      group = UserGroup.where(name: item[:name]).first_or_create!
+      user_ids = User.where(email: item[:users]).pluck(:id)
+      group.update(user_ids: user_ids)
+    end
+  end
+
+  private def import_collections!
+    collections.each do |item|
+      collection = Collection.where(name: item[:name], collection_type: collection_type(item)).first_or_create!
+      collection_relations.each_key do |relation|
+        names = item[relation].select { |m| m[:found] }.map { |r| r[:name] }
+        case relation
+        when :coc_codes
+          collection.update(coc_codes: names) if names.present?
+        else
+          ids = collection.class_name_for_viewable_type(relation).constantize.where(name: names).pluck(:id)
+          collection.set_viewables({ relation => ids }) if ids.present?
+        end
+      end
+    end
+  end
+
+  private def import_access_controls!
+    # Make sure we know about roles, collections, and user groups we created earlier
+    clear_cache
+    access_controls.select { |acl| acl[:existing_user_id].blank? }.each do |item|
+      next if item[:error]
+
+      role = existing_role(item[:role])
+      collection = existing_collection(item[:collection])
+      user_group = existing_user_group(item[:user_group])
+      AccessControl.create!(role: role, collection: collection, user_group: user_group)
+    end
+  end
+
+  private def titles
+    @titles ||= [].tap do |h|
+      Role.permissions(exclude_health: true).count.times do |i|
+        title = roles_worksheet[1][2 + i]&.value
+        h << title if title.present?
+      end
+    end
+  end
+
+  private def permissions_for_row(row, existing_role)
+    {}.tap do |p|
+      Role.permissions(exclude_health: true).count.times do |i|
+        title = titles[i]
+        value = case row[2 + i]&.value
+        when 'X', 'x', 'Y', 'y', 'TRUE', 'true'
+          true
+        else
+          false
```

Comment:
maybe check for 'negative' values too and raise if we get nonsense
```suggestion
        when /\A(?:n|no|false)\z/
          false
        when nil
          false
        else
          # or track this as an error somehow
          raise ArgumentError, "Unknown value: #{row[2 + i]&.value}"
        end
```